src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#include "async-context.h"
+#include "jsg.h"
+#include "setup.h"
+#include <v8.h>
+
+namespace workerd::jsg {
+
+namespace {
+kj::Maybe<AsyncContextFrame&> tryUnwrapFrame(v8::Isolate* isolate, v8::Local<v8::Value> handle) {
+  // Gets the held AsyncContextFrame from the opaque wrappable but does not consume it.
+  KJ_IF_MAYBE(wrappable, Wrappable::tryUnwrapOpaque(isolate, handle)) {
+    OpaqueWrappable<kj::Own<AsyncContextFrame>>* holder =
+        dynamic_cast<OpaqueWrappable<kj::Own<AsyncContextFrame>>*>(wrappable);
+    KJ_ASSERT(holder != nullptr);
+    KJ_ASSERT(!holder->movedAway);
+    return *holder->value;
+  }
+  return nullptr;
+}
+
+}  // namespace
+
+AsyncContextFrame::AsyncContextFrame(IsolateBase& isolate)
+    : isolate(isolate) {}
+
+AsyncContextFrame::AsyncContextFrame(
+    Lock& js,
+    kj::Maybe<AsyncContextFrame&> maybeParent,
+    kj::Maybe<StorageEntry> maybeStorageEntry)
+    : AsyncContextFrame(IsolateBase::from(js.v8Isolate)) {
+  KJ_IF_MAYBE(parent, maybeParent) {
+    parent->propagateTo(js, this->storage, kj::mv(maybeStorageEntry));
+  } else {
+    current(js).propagateTo(js, this->storage, kj::mv(maybeStorageEntry));
+  }
+}
+
+void AsyncContextFrame::propagateTo(
+    Lock& js,
+    Storage& other,
+    kj::Maybe<StorageEntry> maybeStorageEntry) {
+  storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
```

Comment:
I guess this method is only ever called from the constructor of the frame holding `storage`, therefore `storage` is always empty and `eraseAll()` here is redundant (albeit not expensive).

The code might be a little clearer if we actually merged this into the constructor above... unless there's somewhere else propagateTo() might be called in the future?

Reply:
No, take another look. This is called on the `current` frame to propagate it's storage context to the newly created one.

Reply:
Ah, yes, sorry, I misread `storage` as referring to `other`.

I still think it's a little disorienting that the constructor for `AsyncContextFrame` is implemented in terms of this `propagateTo()` method that is oriented in the opposite direction.